include/zephyr/drivers/auxdisplay.h

```diff
@@ -0,0 +1,887 @@
+ *			size character width by character height, values should be 0x00 for
+ *			pixel off or 0xff for pixel on, if a display supports shades then
+ *			values between 0x00 and 0xff may be used (display driver dependent).
+ *
+ * @param dev		Auxiliary display device instance
+ * @param character	Pointer to custom character structure
+ *
+ * @retval		0 on success.
+ * @retval		-ENOSYS if not supported/implemented.
+ * @retval		-EINVAL if provided argument is invalid.
+ * @retval		-EIO if powered down.
+ * @retval		-errno Negative errno code on other failure.
+ */
+__syscall int auxdisplay_custom_character_set(const struct device *dev,
+					      struct auxdisplay_character *character);
+
+static inline int z_impl_auxdisplay_custom_character_set(const struct device *dev,
+							 struct auxdisplay_character *character)
+{
+	struct auxdisplay_driver_api *api = (struct auxdisplay_driver_api *)dev->api;
+
+	if (!api->custom_character_set) {
+		return -ENOSYS;
+	}
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	pm_device_state_get(dev, &state);
+
+	if (state == PM_DEVICE_STATE_SUSPENDED) {
+		return -EIO;
+	}
+#endif
+
+	return api->custom_character_set(dev, character);
+}
+
+/**
+ * @brief	Write data to auxiliary display screen at current position
+ *
+ * @param dev	Auxiliary display device instance
+ * @param data	Text data to write
+ * @param len	Length of text data to write
+ *
+ * @retval	0 on success.
+ * @retval	-EIO if powered down.
+ * @retval	-EINVAL if provided argument is invalid.
+ * @retval	-errno Negative errno code on other failure.
+ */
+__syscall int auxdisplay_write(const struct device *dev, const uint8_t *data,
+			       uint16_t len);
+
+static inline int z_impl_auxdisplay_write(const struct device *dev,
+					  const uint8_t *data, uint16_t len)
+{
+	struct auxdisplay_driver_api *api = (struct auxdisplay_driver_api *)dev->api;
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	pm_device_state_get(dev, &state);
+
+	if (state == PM_DEVICE_STATE_SUSPENDED) {
+		return -EIO;
+	}
+#endif
+
+	return api->write(dev, data, len);
+}
+
+/**
+ * @brief	Send a custom command to the display (if supported by driver)
+ *
+ * @param dev	Auxiliary display device instance
+ * @param data	Custom command structure (this may be extended by specific drivers)
+ *
+ * @retval	0 on success.
+ * @retval	-ENOSYS if not supported/implemented.
+ * @retval	-EINVAL if provided argument is invalid.
+ * @retval	-EIO if powered down.
+ * @retval	-errno Negative errno code on other failure.
+ */
+__syscall int auxdisplay_custom_command(const struct device *dev,
+					struct auxdisplay_custom_data *data);
+
+static inline int z_impl_auxdisplay_custom_command(const struct device *dev,
+						   struct auxdisplay_custom_data *data)
+{
+	struct auxdisplay_driver_api *api = (struct auxdisplay_driver_api *)dev->api;
+
+	if (!api->custom_command) {
+		return -ENOSYS;
+	}
+
+#ifdef CONFIG_PM_DEVICE
```

Comment:
I'm not  crazy about all of the pm stuff inline, and would also expect the device to just wake up when I/O happens rather than return an error (but I'm not a pm expert)

Reply:
Not how PM works with other systems, if the device is suspended it needs to be resumed before any functions will work, I put the PM code here because no driver is able to do anything if the device is in sleep mode rather than duplicating in every auxdisplay driver but maybe this implementation is wrong, comments @gmarull ?

Reply:
@gmarull can you let me know if this is OK or if it should be changed? This seems like the final thing needed before this can be merged.

Reply:
I'll mark this as resolved, since I honestly do not deal with PM a lot and there is already input from @gmarull